C/tests/c4DatabaseTest.cc

```diff
@@ -947,19 +950,93 @@ TEST_CASE("Database Upgrade From 2.8", "[Database][Upgrade][C]") {
 
     C4Log("---- Opening copy of db %s with flags 0x%x", dbPath.c_str(), withFlags);
     C4DatabaseConfig2 config = {slice(TempDir()), withFlags};
-    C4Database *db;
+    c4::ref<C4Database> db;
     auto name = C4Test::copyFixtureDB(kVersionedFixturesSubDir + dbPath);
     C4Log("---- copy Fixture to: %s/%s", TempDir().c_str(), name.asString().c_str());
     C4Error err;
     db = c4db_openNamed(name, &config, WITH_ERROR(&err));
     CHECK(db);
 
-    Doc info(alloc_slice(c4db_getIndexesInfo(db, WITH_ERROR(&err))));
-    CHECK(info.asArray().count() == 1);
-    Dict index = info.asArray()[0].asDict();
-    REQUIRE(index);
-    CHECK(index["name"].asString() == "index1");
-    CHECK(index["type"].asInt() == kC4ValueIndex);
+    std::function<void(C4Database*)> checkIndex = [](C4Database* db){
+        C4Error err;
+        Doc info(alloc_slice(c4db_getIndexesInfo(db, WITH_ERROR(&err))));
+        CHECK(info.asArray().count() == 1);
+        Dict index = info.asArray()[0].asDict();
+        REQUIRE(index);
+        CHECK(index["name"].asString() == "index1");
+        CHECK(index["type"].asInt() == kC4ValueIndex);
+    };
+    checkIndex(db);
+
+    REQUIRE(c4db_close(db, WITH_ERROR()));
+
+    // Reopen the database and edit the first document
+
+    db = c4db_openNamed(name, &config, WITH_ERROR(&err));
+    CHECK(db);
+    CHECK(c4db_getDocumentCount(db) == 2);
+
+    // Get the first document
+
+    C4EnumeratorOptions options = kC4DefaultEnumeratorOptions;
+    c4::ref<C4DocEnumerator> e = c4db_enumerateAllDocs(db, &options, ERROR_INFO());
+    CHECK(c4enum_next(e, ERROR_INFO(&err)));
+
+    c4::ref<C4Document> doc1 = c4enum_getDocument(e, &err);
+    CHECK(doc1);
+    alloc_slice doc1ID = doc1->docID;
+    alloc_slice revID1 = doc1->revID;
+
+    // Get the revision body
+
+    Doc fl_docRev1(alloc_slice(c4doc_getRevisionBody(doc1)), kFLTrusted, c4db_getFLSharedKeys(db));
+    CHECK(fl_docRev1.root().type() == kFLDict);
+    Dict rev1Dict = fl_docRev1.root().asDict();
+
+    std::function<void(Dict&, std::array<slice,2>, std::array<slice,2>)> checkDict
+        = [](Dict& dict, std::array<slice,2> keys, std::array<slice,2> values) {
+            int i = 0;
+            for (Dict::iterator it = dict.begin(); it != dict.end(); ++it) {
+                if (i >= keys.size()) {
+                    return;
+                }
+                CHECK(it.keyString().asString() == keys[i].asString());
+                CHECK(it.value().asString().asString() == values[i].asString());
+                ++i;
+            }
+        };
+
+    checkDict(rev1Dict, {"firstName"_sl, "lastName"_sl}, {"fName"_sl, "lName"_sl});
```

Comment:
This is testing pure KV reading, and is also extraneous to the test.  Reading from the DB does not involve the index.  The other upgrade tests will simply check the doc IDs and previously this test didn't check reading at all.